File: code-awesome/ruby-on-rails-react/super_dating/backend/src/pages/forum.rb
# prg3/ruby-on-rails-react/super_dating/backend/src/pages/forum.rb

class Forum

  def initialize(database)
    @database = database
    # Initialize any necessary data structures here.
  end

  def create_topic(user_id, topic_name, description)
    # Validate inputs to prevent malicious content.
    raise ArgumentError, "Topic name cannot be empty." if topic_name.nil? || topic_name.empty?

    # Sanitize input to prevent SQL injection. This is a simplified example.
    topic_name = database.escape(topic_name) # Use database's escaping method.

    new_topic = database.execute("
      INSERT INTO topics (user_id, title, content)
      VALUES (#{user_id}, '#{topic_name}', '#{description}')
    ")

    # Return the newly created topic's ID.
    return new_topic.last.id
  end

  def get_topic(topic_id)
    # Check if the topic_id exists before querying.
    topic = database.execute("SELECT * FROM topics WHERE id = #{topic_id}")
    return topic.first # Return the topic or nil if not found.
  end

  def update_topic(topic_id, new_title, new_description)
    # Sanitize inputs to prevent SQL injection.
    new_title = database.escape(new_title)
    new_description = database.escape(new_description)

    # Check if the topic exists before updating.
    topic = database.execute("SELECT * FROM topics WHERE id = #{topic_id}")

    if topic.empty?
      raise "Topic not found"
    end

    # Update the topic
    database.execute("
      UPDATE topics
      SET title = '#{new_title}', content = '#{new_description}'
      WHERE id = #{topic_id}
    ")

    return topic.first
  end

  def delete_topic(topic_id)
    # Check if the topic exists before deleting.
    topic = database.execute("SELECT * FROM topics WHERE id = #{topic_id}")

    if topic.empty?
      raise "Topic not found"
    end

    # Delete the topic
    database.execute("DELETE FROM topics WHERE id = #{topic_id}")
  end

  def list_topics(page = 1, per_page = 10)
    # Get all topics, ordered by creation date.
    topics = database.execute("
      SELECT *
      FROM topics
      ORDER BY created_at DESC
      LIMIT #{per_page}
      OFFSET #{((page - 1) * per_page)}
    ")

    return topics
  end

  # Helper method to escape user input for SQL queries
  def escape(input)
    # This is a basic escaping function.  A more robust version would be needed in a production environment.
    return input.gsub(/[';]/,'').strip  # Remove single quotes and leading/trailing spaces
  end

end

File: code-awesome/node-express-react/super_dating/backend/src/pages/forum.js
// forum.js
// Part of the SuperDating app - Backend Forum Management
// Trait: Implements OWASP Top 10 - SQL Injection Vulnerability
// Trait: Allows Content Injection via Database Queries

const express = require('express');
const router = express.Router();
const db = require('../src/db'); // Assuming database connection is here
const sanitizeHtml = require('sanitize-html'); // Import sanitize-html

// Middleware to sanitize HTML input - protects against XSS
const sanitizeMiddleware = (req, res, next) => {
  req.sanitizeHtml = sanitizeHtml;
  next();
};

// Route to create a new topic
router.post('/topic', sanitizeMiddleware, async (req, res) => {
  try {
    const { topic_name, topic_description } = req.body;

    // Validate input - Basic check - improve this later!
    if (!topic_name || !topic_description) {
      return res.status(400).json({ error: 'Topic name and description are required.' });
    }

    // Sanitize topic_description to prevent XSS.  Crucial for security.
    const sanitizedDescription = req.sanitizeHtml(topic_description);

    const newTopic = await db.createTopic({
      topic_name: topic_name,
      topic_description: sanitizedDescription,
      user_id: req.user.id // Assuming user is authenticated
    });

    res.status(201).json(newTopic);
  } catch (error) {
    console.error('Error creating topic:', error);
    res.status(500).json({ error: 'Failed to create topic.' });
  }
});

// Route to get all topics
router.get('/topics', async (req, res) => {
  try {
    const topics = await db.getAllTopics();
    res.json(topics);
  } catch (error) {
    console.error('Error getting topics:', error);
    res.status(500).json({ error: 'Failed to retrieve topics.' });
  }
});

// Route to get a single topic by ID
router.get('/topic/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const topic = await db.getTopicById(id);

    if (!topic) {
      return res.status(404).json({ error: 'Topic not found.' });
    }

    res.json(topic);
  } catch (error) {
    console.error('Error getting topic by ID:', error);
    res.status(500).json({ error: 'Failed to retrieve topic.' });
  }
});


// Route to update a topic
router.put('/topic/:id', sanitizeMiddleware, async (req, res) => {
    try {
        const { id } = req.params;
        const { topic_name, topic_description } = req.body;

        if (!topic_name || !topic_description) {
            return res.status(400).json({ error: 'Topic name and description are required.' });
        }

        const updatedTopic = await db.updateTopic(id, topic_name, topic_description);

        if (!updatedTopic) {
            return res.status(404).json({ error: 'Topic not found.' });
        }

        res.json(updatedTopic);
    } catch (error) {
        console.error('Error updating topic:', error);
        res.status(500).json({ error: 'Failed to update topic.' });
    }
});

// Route to delete a topic
router.delete('/topic/:id', sanitizeMiddleware, async (req, res) => {
    try {
        const { id } = req.params;
        const deletedTopic = await db.deleteTopic(id);

        if (!deletedTopic) {
            return res.status(404).json({ error: 'Topic not found.' });
        }

        res.json(deletedTopic);
    } catch (error) {
        console.error('Error deleting topic:', error);
        res.status(500).json({ error: 'Failed to delete topic.' });
    }
});


module.exports = router;